test/router/V4Router.t.sol

```diff
@@ -145,45 +98,125 @@ contract V4RouterTest is RoutingTestHelpers {
+
+        _finalizeExecuteAndCheckSwap(nativeKey.currency0, nativeKey.currency1, amountIn, expectedAmountOut);
+    }
+
+    function test_nativeOut_swapExactInputSingle() public {
+        uint256 amountIn = 1 ether;
+        uint256 expectedAmountOut = 992054607780215625;
+
+        // native output means we need !zeroForOne
+        IV4Router.ExactInputSingleParams memory params =
+            IV4Router.ExactInputSingleParams(nativeKey, false, uint128(amountIn), 0, 0, bytes(""));
+
+        plan = plan.add(Actions.SWAP_EXACT_IN_SINGLE, abi.encode(params));
+
+        _finalizeExecuteAndCheckSwap(nativeKey.currency1, nativeKey.currency0, amountIn, expectedAmountOut);
+    }
+
+    function test_nativeIn_swapExactIn_1Hop() public {
+        uint256 amountIn = 1 ether;
+        uint256 expectedAmountOut = 992054607780215625;
+
+        tokenPath.push(CurrencyLibrary.NATIVE);
+        tokenPath.push(currency0);
+        IV4Router.ExactInputParams memory params = _getExactInputParams(tokenPath, amountIn);
+        plan = plan.add(Actions.SWAP_EXACT_IN, abi.encode(params));
+
+        _finalizeExecuteAndCheckSwap(CurrencyLibrary.NATIVE, currency0, amountIn, expectedAmountOut);
+    }
+
+    function test_nativeOut_swapExactIn_1Hop() public {
+        uint256 amountIn = 1 ether;
+        uint256 expectedAmountOut = 992054607780215625;
+
+        tokenPath.push(currency0);
+        tokenPath.push(CurrencyLibrary.NATIVE);
+        IV4Router.ExactInputParams memory params = _getExactInputParams(tokenPath, amountIn);
 
         plan = plan.add(Actions.SWAP_EXACT_IN, abi.encode(params));
-        bytes memory data = plan.finalizeSwap(currency0, currency3, address(this));
 
-        router.executeActions(data);
+        _finalizeExecuteAndCheckSwap(currency0, CurrencyLibrary.NATIVE, amountIn, expectedAmountOut);
+    }
 
-        uint256 newBalance0 = currency0.balanceOfSelf();
-        uint256 newBalance3 = currency3.balanceOfSelf();
+    function test_nativeIn_swapExactIn_2Hops() public {
+        uint256 amountIn = 1 ether;
+        uint256 expectedAmountOut = 984211133872795298;
+
+        tokenPath.push(CurrencyLibrary.NATIVE);
+        tokenPath.push(currency0);
+        tokenPath.push(currency1);
+        IV4Router.ExactInputParams memory params = _getExactInputParams(tokenPath, amountIn);
+
+        plan = plan.add(Actions.SWAP_EXACT_IN, abi.encode(params));
+
+        uint256 intermediateBalanceBefore = currency0.balanceOfSelf();
+
+        _finalizeExecuteAndCheckSwap(CurrencyLibrary.NATIVE, currency1, amountIn, expectedAmountOut);
 
-        assertEq(prevBalance0 - newBalance0, amountIn);
-        assertEq(newBalance3 - prevBalance3, expectedAmountOut);
+        // check intermediate token balances
+        uint256 intermediateBalanceAfter = currency0.balanceOfSelf();
+        assertEq(intermediateBalanceBefore - intermediateBalanceAfter, 0);
         assertEq(currency0.balanceOf(address(router)), 0);
```

Comment:
i thought balance checks of router were also checked in _finalizeExecuteAndCheck swap?

Reply:
only the input and output token balances are checked so these tests still check intermediate tokens